consensus/polybft/consensus_runtime.go

```diff
@@ -222,26 +239,35 @@ func (c *consensusRuntime) OnBlockInserted(block *types.Block) {
 		c.logger.Error("on block inserted error", "err", err)
 	}
 
+	var (
+		proposerSnapshot = c.proposerSnapshot.Copy()
+		epoch            = c.epoch
+		err              error
+	)
+
 	if c.isEndOfEpoch(block.Header.Number) {
-		// reset the epoch. Internally it updates the parent block header.
-		if err := c.restartEpoch(block.Header); err != nil {
+		if epoch, err = c.restartEpoch(block.Header); err != nil {
 			c.logger.Error("failed to restart epoch after block inserted", "error", err)
-		}
-	} else {
-		c.lock.Lock()
-		c.lastBuiltBlock = block.Header
 
-		proposerSnapshot := c.proposerSnapshot.Copy()
-		if err := c.updateProposerSnapshot(proposerSnapshot, block.Number()); err != nil {
-			c.logger.Warn("Could not update proposer snapshot", "err", err)
+			return
 		}
+	}
+
+	if err := c.updateProposerSnapshot(proposerSnapshot, block.Number()); err != nil {
+		c.logger.Warn("Could not update proposer calculator", "err", err)
 
-		c.proposerSnapshot = proposerSnapshot
-		c.lock.Unlock()
+		return
 	}
+
+	// finally update runtime state (lastBuiltBlock, epoch, proposerSnapshot)
+	c.epoch = epoch
+	c.lastBuiltBlock = block.Header
+	c.proposerSnapshot = proposerSnapshot
 }
 
 func (c *consensusRuntime) createCommitmentAndBundles(txs []*types.Transaction) error {
+	parentBlock, epoch := c.lastBuiltBlock, c.epoch
```

Comment:
Is this safe to do this way, instead of acquiring a copy via `getLastBuiltBlockAndEpoch`?

Reply:
this method is called under the lock